x-pack/metricbeat/module/containerd/cpu/cpu.go

```diff
@@ -0,0 +1,234 @@
+	preContainerCpuTotalUsage  map[string]float64
+	preContainerCpuKernelUsage map[string]float64
+	preContainerCpuUserUsage   map[string]float64
+}
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
+	mb.Registry.MustAddMetricSet("containerd", "cpu", New,
+		mb.WithHostParser(hostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The containerd cpu metricset is beta.")
+
+	pc, err := prometheus.NewPrometheusClient(base)
+	if err != nil {
+		return nil, err
+	}
+	config := containerd.DefaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+	mod, ok := base.Module().(containerd.Module)
+	if !ok {
+		return nil, fmt.Errorf("must be child of kubernetes module")
+	}
+	return &metricset{
+		BaseMetricSet:              base,
+		prometheusClient:           pc,
+		mod:                        mod,
+		calcPct:                    config.CalculateCpuPct,
+		preTimestamp:               time.Time{},
+		preContainerCpuTotalUsage:  map[string]float64{},
+		preContainerCpuKernelUsage: map[string]float64{},
+		preContainerCpuUserUsage:   map[string]float64{},
+	}, nil
+}
+
+// Fetch gathers information from the containerd and reports events with this information.
+func (m *metricset) Fetch(reporter mb.ReporterV2) error {
+	families, timestamp, err := m.mod.GetContainerdMetricsFamilies(m.prometheusClient)
+	if err != nil {
+		return errors.Wrap(err, "error getting families")
+	}
+
+	events, err := m.prometheusClient.ProcessMetrics(families, mapping)
+	if err != nil {
+		return errors.Wrap(err, "error getting events")
+	}
+
+	perContainerCpus := make(map[string]int)
+	if m.calcPct {
+		for _, event := range events {
+			if _, err = event.GetValue("cpu"); err == nil {
+				// calculate cpus used by each container
+				setContCpus(event, perContainerCpus)
+			}
+		}
+	}
+
+	for _, event := range events {
+		// setting ECS container.id and module field containerd.namespace
+		containerFields := common.MapStr{}
+		moduleFields := common.MapStr{}
+		rootFields := common.MapStr{}
+
+		cID := containerd.GetAndDeleteCid(event)
+		namespace := containerd.GetAndDeleteNamespace(event)
+
+		containerFields.Put("id", cID)
+		rootFields.Put("container", containerFields)
+		moduleFields.Put("namespace", namespace)
+
+		if m.calcPct {
+			contCpus, ok := perContainerCpus[cID]
+			if !ok {
+				contCpus = 1
+			}
+			// calculate timestamp delta
+			timestampDelta := int64(0)
+			if !m.preTimestamp.IsZero() {
+				timestampDelta = timestamp.UnixNano() - m.preTimestamp.UnixNano()
+			}
+			// Calculate cpu total usage percentage
+			cpuUsageTotal, err := event.GetValue("usage.total.ns")
+			if err == nil {
```

Comment:
In this case, we iterate a batch of events and we check if `usage.total.ns` field is present. There is for sure one event in this batch that this field is not present and that is the one that has `system.total` field. The reason is that `process_cpu_seconds_total` containerd metric that is then mapped to `system.total`  field, does not include info about any container id (system wise metric not container specific). So that leads to an event that has only `system.total` field.  While the rest of the events have fields that are grouped together due to `container_id` mainly. So with this `if` I just want to be sure that we skip trying to calculate percentages for the event that only has `system.total` (it does not include any other field, is not container specific). It is not an actual error to log anything. But I will add a comment.